### ISSUE TITLE
await block results in "cannot return value referencing local variable" for Copy type

### ISSUE THREAD

**Issue by alex** [rust-lang/rust#67785]
Stripped down code [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9992414cc911aa7bd4e781aea65171e2):

```rust
async fn submit_to_log() -> Result<(), ()> {
    Ok(())
}

pub async fn submit_cert_to_logs(logs: &[()]) -> Vec<Option<usize>> {
    let futures = logs
        .iter()
        .enumerate()
        .map(move |(idx, ())| {
            let s = submit_to_log();
            async {
                match s.await {
                    Ok(()) => Some(idx),
                    _ => None,
                }
            }
        })
        .collect::<Vec<_>>();

    futures::future::join_all(futures).await
}
```

This complains that `cannot return value referencing local variable idx`, however `idx` is a `usize`, which means it's `Copy`. As far as I understand it, that means the compiler can "solve" this problem for me by simply copying the value implicitly.

**Comment by jonas-schievink** [comment]
`async move` fixes this. This seems to be expected behavior, similar to how closures work.

**Comment by alex** [comment]
Why yes, yes it does :-) Thanks for the super quick answer.

I'm not totally I understand why this is expected behavior though, and either way I suspect there may be a way to turn this into improved diagnostics.
